File: agents/issue-reconciler/fingerprint.ts
// Shared by board-history (parses fingerprints left by earlier runs) and the
// writer (stamps a new one on every comment). Keeping both in one place
// means the format can never drift between reader and writer.
// Format: <!-- issue-reconciler: v1 run=<id> decision=<action> evidence=sha256:<hex> -->

const FINGERPRINT_RE = /<!--\s*issue-reconciler:\s*v1\s+run=(\S+)\s+decision=(\S+)\s+evidence=sha256:([0-9a-f]+)\s*-->/;

export type Fingerprint = {
  runId: string;
  decision: string;
  evidenceHash: string;
};

export function buildFingerprint(fp: Fingerprint): string {
  return `<!-- issue-reconciler: v1 run=${fp.runId} decision=${fp.decision} evidence=sha256:${fp.evidenceHash} -->`;
}

export function parseFingerprint(commentBody: string): Fingerprint | null {
  const match = FINGERPRINT_RE.exec(commentBody);
  if (!match) return null;
  const [, runId, decision, evidenceHash] = match;
  return { runId, decision, evidenceHash };
}

File: agents/issue-reconciler/issue_reconciler/fingerprint.py
"""Shared by board_history (parses fingerprints left by earlier runs) and
the comment writer (stamps a new one on every comment). Keeping both in one
place means the format can never drift between reader and writer.
Format: <!-- issue-reconciler: v1 run=<id> decision=<action> evidence=sha256:<hex> -->
"""
from __future__ import annotations

import re
from typing import TypedDict

_FINGERPRINT_RE = re.compile(
    r"<!--\s*issue-reconciler:\s*v1\s+run=(\S+)\s+decision=(\S+)\s+evidence=sha256:([0-9a-f]+)\s*-->"
)


class Fingerprint(TypedDict):
    run_id: str
    decision: str
    evidence_hash: str


def build_fingerprint(fp: Fingerprint) -> str:
    return f"<!-- issue-reconciler: v1 run={fp['run_id']} decision={fp['decision']} evidence=sha256:{fp['evidence_hash']} -->"


def parse_fingerprint(comment_body: str) -> Fingerprint | None:
    match = _FINGERPRINT_RE.search(comment_body)
    if not match:
        return None
    run_id, decision, evidence_hash = match.groups()
    return {"run_id": run_id, "decision": decision, "evidence_hash": evidence_hash}
